### ISSUE TITLE
prefer-immutable-types: incorrectly inferring ReadonlyShallow array as ReadonlyDeep

### ISSUE THREAD

**Issue by danielnixon** [eslint-functional/eslint-plugin-functional#578]
## Bug Report

```ts
type Thing = { a: string };

type OtherThing = { readonly a: string };

declare const a: readonly OtherThing[];

// Surprising. I would expect ReadonlyShallow, not ReadonlyDeep
// Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep"). eslint functional/prefer-immutable-types
const b: readonly Thing[] = a;
```

### Expected behavior

`const b: readonly Thing[] = a;` is identified as ReadonlyShallow

### Actual behavior

`const b: readonly Thing[] = a;` is identified as ReadonlyDeep

### Steps to reproduce

See example above.

**Comment by danielnixon** [comment]
This has some heisenbug properties. Commenting out an unrelated (?) line affects the result.

```ts
type Thing = { a: string };

type OtherThing = { readonly a: string };

declare const a: readonly OtherThing[];

// Variable should have an immutability of at least "Immutable" (actual: "ReadonlyDeep")
declare const c: readonly Thing[];
```

versus

```ts
type Thing = { a: string };

type OtherThing = { readonly a: string };

// declare const a: readonly OtherThing[];

// Variable should have an immutability of at least "Immutable" (actual: "ReadonlyShallow")
declare const c: readonly Thing[];
```

**Comment by RebeccaStevens** [comment]
Is this still an issue?

**Comment by RebeccaStevens** [comment]
I finally got around to testing this.
I can't reproduce this issue so this hopefully got fixed at some point.
I'm getting an immutability result of `ReadonlyShallow` in both cases.
